src/ORM/FieldType/DBGeneratedColumn.php

```diff
@@ -0,0 +1,288 @@
+        DBGeneratedColumn::class,
+        DBPolymorphicForeignKey::class,
+        DBPrimaryKey::class,
+    ];
+
+    private DBField $childField;
+
+    private string $generationExpression;
+
+    private string $generationType;
+
+    public function __construct(?string $name = null, string $fieldSpec = '', string $expression = '', string $type = DBGeneratedColumn::GENERATION_VIRTUAL)
+    {
+        if (!in_array($type, [DBGeneratedColumn::GENERATION_STORED, DBGeneratedColumn::GENERATION_VIRTUAL])) {
+            throw new InvalidArgumentException('$type must be "STORED" or "VIRTUAL"');
+        }
+        if ($name === null && $fieldSpec === '' && $expression === '') {
+            // DataObjectSchema needs to be able to make a singleton with no args.
+            // This means we need these to be able to be empty....
+            // In that case, we can just skip all the fancy stuff.
+            // Don't use the parent constructor because it calls setValue() even though we have no value.
+            return;
+        } elseif ($fieldSpec === '' || $expression === '') {
+            throw new InvalidArgumentException('$fieldSpec and $expression must not be empty');
+        }
+        $this->generationExpression = $expression;
+        $this->generationType = $type;
+        $this->childField = DBField::create_field($fieldSpec, null, $name);
+        $this->validateChildFieldClass();
+        $this->setFailover($this->childField);
+        parent::__construct($name, []);
+    }
+
+    public function getChildField(): DBField
+    {
+        return $this->childField;
+    }
+
+    public function requireField(): void
+    {
+        $spec = $this->getChildField()->getFieldSpec();
+        if (is_string($spec)) {
+            $defaultValue = $this->getChildField()->getDefaultValue();
+            $spec = DB::get_schema()->makeGenerated(
+                $spec,
+                ['default' => $defaultValue],
+                $this->generationExpression,
+                $this->generationType
+            );
+        } else {
+            $spec['generated'] = [
+                'type' => $this->generationType,
+                'expression' => $this->generationExpression,
+            ];
+        }
+        DB::require_field($this->getTable(), $this->getName(), $spec);
+    }
+
+    public function scaffoldFormField(?string $title = null, array $params = []): ?FormField
+    {
+        return $this->getChildField()->scaffoldFormField($title, $params)->performReadonlyTransformation();
+    }
+
+    public function scaffoldSearchField(?string $title = null): ?FormField
+    {
+        return $this->getChildField()->scaffoldSearchField($title);
+    }
+
+    public function defaultSearchFilter(?string $name = null): SearchFilter
+    {
+        return $this->getChildField()->defaultSearchFilter($name);
+    }
+
+    public function getIndexSpecs(): ?array
+    {
+        return $this->getChildField()?->getIndexSpecs();
```

Comment:
```suggestion
        return $this->getChildField()->getIndexSpecs();
```
Match the others methods, or remove the shorthand and do and if statement and inline comment explaining scenario where it could be null

Reply:
IIRC this one has to be there or else the singleton throws an exception. Will double check and add a comment if it is needed.

I don't see the need to change this to a conditional statement, will keep as is if it's needed.

Reply:
Doesn't seem to be needed anymore.